### PR TITLE
Support canonical name of python objects

### PR DESCRIPTION
**Pull request by tk0miya**
### Feature or Bugfix
- Feature

### Purpose
- We can give some names to object in python. But current implementation only supports to describe a public name. But a runtime object only knows the name where the object is defined. So supporting canonical name for the object is important for autodoc.
- This allows to describe additional name as a canonical name.
- refs: #4826, #4944

### CODE REVIEW

File: sphinx/domains/python.py [hunk 1]
```diff
@@ -69,7 +69,8 @@
 
 ObjectEntry = NamedTuple('ObjectEntry', [('docname', str),
                                          ('node_id', str),
-                                         ('objtype', str)])
+                                         ('objtype', str),
+                                         ('canonical', bool)])
```
Comment (tk0miya) [inline sphinx/domains/python.py:69]: This changes the internal structure of python domain. So this might be an incompatible change....

**Review by shimizukawa — APPROVED** [review]
LGTM!

File: sphinx/domains/python.py (hunk 1 above)
Comment (tk0miya) [reply sphinx/domains/python.py:69]: Finally, I and @shimizukawa considered this is an incompatible change of Python domain's API. So this should be merged into 4.0 release.

Comment (shimizukawa) [reply sphinx/domains/python.py:69]: 👍

**Pull request merged**